src/psyclone/transformations.py

```diff
@@ -2840,101 +2863,89 @@ def name(self):
 
-    def apply(self, kern, options=None):
+    def apply(self, node, options=None):
         '''
-        Modifies the AST of the supplied kernel so that it contains an
-        '!$acc routine' OpenACC directive.
-        This transformation affects the f2pygen and the PSyIR trees of
-        this kernel.
+        Add the '!$acc routine' OpenACC directive into the code of the
+        supplied kernel or routine.
 
-        :param kern: the kernel object to transform.
-        :type kern: :py:class:`psyclone.psyGen.Kern`
+        :param node: the kernel object to transform.
+        :type node: :py:class:`psyclone.psyGen.Kern` or \
+                    :py:class:`psyclone.psyir.nodes.Routine`
         :param options: a dictionary with options for transformations.
         :type options: dictionary of string:values or None
 
-        :raises TransformationError: if we fail to find the subroutine \
-                                     corresponding to the kernel object.
-
         '''
-        # pylint: disable=too-many-locals
-
         # Check that we can safely apply this transformation
-        self.validate(kern, options)
-
-        # Get the fparser2 AST of the kernel
-        ast = kern.ast
-        # Find the kernel subroutine in the fparser2 parse tree
-        kern_sub = None
-        subroutines = walk(ast.content, Subroutine_Subprogram)
-        for sub in subroutines:
-            for child in sub.content:
-                if isinstance(child, Subroutine_Stmt) and \
-                   str(child.items[1]) == kern.name:
-                    kern_sub = sub
-                    break
-            if kern_sub:
-                break
-        # Find the last declaration statement in the subroutine
-        spec = walk(kern_sub.content, Specification_Part)[0]
-        posn = -1
-        for idx, node in enumerate(spec.content):
-            if not isinstance(node, (Implicit_Part, Type_Declaration_Stmt)):
-                posn = idx
-                break
-        # Create the directive and insert it
-        cmt = Comment(FortranStringReader("!$acc routine",
-                                          ignore_comments=False))
-        if posn == -1:
-            spec.content.append(cmt)
-        else:
-            spec.content.insert(posn, cmt)
+        self.validate(node, options)
 
-        # Flag that the kernel has been modified
-        kern.modified = True
+        if isinstance(node, Kern):
+            # Flag that the kernel has been modified
+            node.modified = True
+
+            # Get the schedule representing the kernel subroutine
+            routine = node.get_kernel_schedule()
+        else:
+            routine = node
 
-        # Add the 'cmt' directive into the PSyIR as a CodeBlock
-        kernel_schedule = kern.get_kernel_schedule()
-        kernel_schedule.addchild(
-            CodeBlock([cmt], CodeBlock.Structure.STATEMENT), 0)
+        # Insert the directive to the routine if it doesn't already exist
```

Comment:
This is so much nicer :-)
Contrary to what I said in an earlier comment, I think we can remove that test that the directive gets added at the end of the declarations - it's not an issue anymore.

Reply:
Done